src/sub.rs

```diff
@@ -8,17 +8,102 @@ use crate::{
     MultiPeerBackend, Socket, SocketBackend, SocketEvent, SocketOptions, SocketRecv, SocketType,
 };
 
-use crate::backend::GenericSocketBackend;
+use crate::backend::Peer;
+use dashmap::DashMap;
+use parking_lot::Mutex;
+use crate::fair_queue::QueueInner;
 use crate::fair_queue::FairQueue;
+use crossbeam::queue::SegQueue;
 use async_trait::async_trait;
 use bytes::{BufMut, BytesMut};
 use futures::channel::mpsc;
 use futures::{SinkExt, StreamExt};
-use std::collections::HashMap;
+use std::collections::{HashMap, HashSet};
 use std::sync::Arc;
 
+pub(crate) struct SubSocketBackend {
+    pub(crate) peers: DashMap<PeerIdentity, Peer>,
+    fair_queue_inner: Option<Arc<Mutex<QueueInner<ZmqFramedRead, PeerIdentity>>>>,
+    pub(crate) round_robin: SegQueue<PeerIdentity>,
+    socket_type: SocketType,
+    socket_options: SocketOptions,
+    pub(crate) socket_monitor: Mutex<Option<mpsc::Sender<SocketEvent>>>,
+    subs: Mutex<HashSet<String>>,
+}
+
+impl SubSocketBackend {
+    pub(crate) fn with_options(
+        fair_queue_inner: Option<Arc<Mutex<QueueInner<ZmqFramedRead, PeerIdentity>>>>,
+        socket_type: SocketType,
+        options: SocketOptions,
+    ) -> Self {
+        Self {
+            peers: DashMap::new(),
+            fair_queue_inner,
+            round_robin: SegQueue::new(),
+            socket_type,
+            socket_options: options,
+            socket_monitor: Mutex::new(None),
+	    subs: Mutex::new(HashSet::new()),
+        }
+    }
+}
+
+impl SocketBackend for SubSocketBackend {
+    fn socket_type(&self) -> SocketType {
+        self.socket_type
+    }
+
+    fn socket_options(&self) -> &SocketOptions {
+        &self.socket_options
+    }
+
+    fn shutdown(&self) {
+        self.peers.clear();
+    }
+
+    fn monitor(&self) -> &Mutex<Option<mpsc::Sender<SocketEvent>>> {
+        &self.socket_monitor
+    }
+}
+
+#[async_trait]
+impl MultiPeerBackend for SubSocketBackend {
+    async fn peer_connected(self: Arc<Self>, peer_id: &PeerIdentity, io: FramedIo) {
+	println!("Connect peer");
+        let (recv_queue, send_queue) = io.into_parts();
+        self.peers.insert(peer_id.clone(), Peer { send_queue });
+        self.round_robin.push(peer_id.clone());
+        match &self.fair_queue_inner {
+            None => {}
+            Some(inner) => {
+                inner.lock().insert(peer_id.clone(), recv_queue);
+            }
+        };
+
+	let subs = (|| { (*self.subs.lock()).clone() })();
+
+	for sub in subs.iter() {
+	    let mut buf = BytesMut::with_capacity(sub.len() + 1);
+	    buf.put_u8(1);
+	    buf.extend_from_slice(sub.as_bytes());
+
+	    let message: ZmqMessage = ZmqMessage::from(buf.freeze());
+	    let mut peer = self.peers.get_mut(&peer_id).unwrap();
```

Comment:
you can use io: FramedIo parameter to send messages directly to this peer. There is no need to do a lookup for it